Kubeadm - Added initial support for Windows worker nodes to join cluster using kubeadm

bsteciuk opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
This PR adds initial support for adding a Windows worker node to a Kubernetes cluster with kubeadm.  Also adds Windows build of kubeadm to node build targets. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes https://github.com/kubernetes/kubeadm/issues/364 

**Special notes for your reviewer**:

Depends on #53730 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
kubeadm: Add support for adding a Windows node
```

brendandburns commented:
/ok-to-test

luxas commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go:
```diff
@@ -94,7 +96,11 @@ func SetDefaults_MasterConfiguration(obj *MasterConfiguration) {
 func SetDefaults_NodeConfiguration(obj *NodeConfiguration) {
 	if obj.CACertPath == "" {
-		obj.CACertPath = DefaultCACertPath
+		if runtime.GOOS == "windows" {
```
Please don't put `if windows { } else { }` generally in the code here. Use generic path helper funcs where there is a difference between platforms

luxas commented on test/e2e_node/system/types.go:
```diff
@@ -164,3 +164,20 @@ var DefaultSysSpec = SysSpec{
 	},
 }
+
+var WindowsSysSpec = SysSpec{
```
could we have a map or something that maps `GOOS` and `SysSpec` possibly?
That would make this more reusable I guess...

luxas commented on test/e2e_node/system/kernel_validator_helper.go:
```diff
@@ -0,0 +1,42 @@
+	"os/exec"
+)
+
+type KernelValidatorHelper interface {
```
Godoc comments everywhere please

luxas commented on test/e2e_node/system/kernel_validator_helper.go:
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017 :)

luxas commented on test/e2e_node/system/kernel_validator.go:
```diff
@@ -61,14 +62,23 @@ const (
 	var errs []error
 	errs = append(errs, k.validateKernelVersion(spec.KernelSpec))
-	errs = append(errs, k.validateKernelConfig(spec.KernelSpec))
+	//only validate kernel config when necessary (currently no kernel config for windows)
```
nit: `// Foo bar baz`

luxas commented on test/e2e_node/system/docker_validator.go:
```diff
@@ -38,6 +39,7 @@ func (d *DockerValidator) Name() string {
 
 const (
 	dockerEndpoint            = "unix:///var/run/docker.sock"
+	windowsDockerEndpoint     = "npipe:////./pipe/docker_engine"
```
Can this be done as a map to avoid the `if else` statement please?

luxas commented on pkg/util/initsystem/initsystem.go:
```diff
@@ -86,7 +86,51 @@ func (sysd SystemdInitSystem) ServiceIsActive(service string) bool {
-// getInitSystem returns an InitSystem for the current system, or nil
+type WindowsInitSystem struct{}
+
+func (sysd WindowsInitSystem) ServiceStart(service string) error {
```
Godoc comments

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -712,26 +748,30 @@ func RunJoinNodeChecks(cfg *kubeadmapi.NodeConfiguration) error {
-		InPathCheck{executable: "touch", mandatory: false},
+	}
+	//non-windows checks
+	if runtime.GOOS != "windows" {
```
can as well do `if linux {`, that's more correct I guess

luxas commented on cmd/kubeadm/app/preflight/checks.go:
Any other method of checking admin privs that might be more obvious?

luxas commented on cmd/kubeadm/app/preflight/checks.go:
windowsPrivilegedCheck() and linuxPrivilegedCheck() possibly?

luxas commented on cmd/kubeadm/app/preflight/checks.go:
can we dedupe and make this code flow a bit cleaner?

luxas reviewed (commented):
Thanks! Great progress, I'm so happy to see you working on this :smile: :+1: 

Left a couple of comments. Mostly around code ordering / etc.

I want you to split all `test/e2e_node` changes into a separate, generic PR for review by sig-node though, and then merge this after that one.

bsteciuk commented:
/retest

bsteciuk committed:
Kubeadm - Added initial support for Windows worker nodes to join cluster using kubeadm

Added kubeadm to node build targets

Created unix/windows specific conditionally compiled for checks.go and defaults.go

bsteciuk commented:
Created #53730 for test/e2e_node changes as requested.

kad commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go:
```diff
@@ -46,7 +47,7 @@ const (
 )
 
-func addDefaultingFuncs(scheme *runtime.Scheme) error {
+func addDefaultingFuncs(scheme *kruntime.Scheme) error {
```
nit: if std "runtime" not used in this file, do we need now to rename that import ?

bsteciuk replied:
yup, runtime had been in there, but is no longer, updating

kad commented on cmd/kubeadm/app/preflight/checks.go:
What do you think about splitting this into more Go idiomatic way to `checks_windows.go` and have it build only in case of windows ?

bsteciuk replied:
:+1:

kad commented on cmd/kubeadm/app/util/path.go:
```diff
@@ -0,0 +1,36 @@
+// on linux, the call is a no-op
+func NormalizePath(path string) string {
+	if runtime.GOOS == "windows" {
+		if filepath.VolumeName(path) == "" {
```
same thing: maybe split it to path_windows.go and build only on windows ?

bsteciuk replied:
@kad would it make more sense just to split up defaults.go for necessary OS specific defaults (at the moment just one path), and get rid of the NormalizePath() func?

kad replied:
I don't know. probably @luxas has opinion about that ?

brendandburns commented on pkg/util/initsystem/initsystem.go:
```diff
@@ -43,13 +43,13 @@ type SystemdInitSystem struct{}
 func (sysd SystemdInitSystem) ServiceStart(service string) error {
 	args := []string{"start", service}
-	_, err := exec.Command("systemctl", args...).Output()
+	err := exec.Command("systemctl", args...).Run()
```
Not going to make you do this, since this was already this way, but it would probably be good to switch to `util.exec` for testability...

luxas replied:
maybe @bsteciuk can follow up later?

brendandburns commented on test/e2e_node/system/kernel_validator.go:
```diff
@@ -46,6 +47,11 @@ func (k *KernelValidator) Name() string {
 	return "kernel"
 }
 
+var kernelValidatorHelpersByOS = map[string]KernelValidatorHelper{
```
probably worth doing the `*_windows.go` here too instead of a function pointer.

bsteciuk replied:
@brendanburns yup, this was already done in the other PR #53730 made for the e2e_node portions.

brendandburns commented:
Awesome to see this start to come together. Some minor comments.

luxas commented:
/hold

^ means we're waiting for https://github.com/kubernetes/kubernetes/pull/53730 to get merged first

brendandburns commented:
/lgtm
/approve

Approved #53730, I'll lift the hold once it merges.

bsteciuk commented:
/retest

bsteciuk commented:
/retest

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -435,15 +426,20 @@ func (sysver SystemVerificationCheck) Check() (warnings, errors []error) {
 	// Run the system verification check, but write to out buffered writer instead of stdout
 	bufw := bufio.NewWriterSize(os.Stdout, 1*1024*1024)
 	reporter := &system.StreamReporter{WriteStream: bufw}
+	runtimeOS := runtime.GOOS
```
I'm not sure why you made this extra variable, things seem to be fine without it (using `runtime.GOOS` directly in this func)

bsteciuk replied:
fixed

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -435,15 +426,20 @@ func (sysver SystemVerificationCheck) Check() (warnings, errors []error) {
-		&system.DockerValidator{Reporter: reporter},
+		&system.DockerValidator{Reporter: reporter}}
+
+	if runtimeOS != "windows" {
```
Can we switch this to `runtime.GOOS == "linux"` instead?

bsteciuk replied:
fixed

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -435,15 +426,20 @@ func (sysver SystemVerificationCheck) Check() (warnings, errors []error) {
+	if runtimeOS != "windows" {
+		//add non-windows validators
+		validators = append(validators,
+			&system.OSValidator{Reporter: reporter},
```
I would expect the system verification check to support windows in this validator; can you fix that?

bsteciuk replied:
@luxas, I'm not sure about this one.  The current check runs `uname` and verifies that it returns `sysSpec.OS`.  For unix, it verifies it returns `Linux` as the unix variant, as opposed to SunOS, or FreeBSD, etc.  I'm not sure there's an equivalent command or even the same "variant" concept in Windows.

luxas commented on cmd/kubeadm/app/preflight/checks_windows.go:
```diff
@@ -0,0 +1,42 @@
+
+package preflight
+
+import "fmt"
```
follow the normal
```go
import (
    "fmt"
    "strings"
...
)
```
pattern please

bsteciuk replied:
fixed.

luxas reviewed (commented):
Thanks @bsteciuk.
I left some nits only, soon this is ready to be merged!

luxas reviewed (approved):
@bsteciuk Only a commit squash away from LGTM ^^

luxas reviewed (approved):
/lgtm
/retest

luxas commented:
/retest

luxas commented:
/retest

luxas commented:
/retest

luxas commented:
flakes on https://github.com/kubernetes/kubernetes/issues/54873

bsteciuk commented:
/retest

krzyzacy commented:
/retest
